mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -607,16 +656,30 @@ def compute_next_token(current_size):
                 query = query.join(j)
 
             offset = SearchUtils.parse_start_offset_from_page_token(page_token)
-            queried_runs = query.distinct() \
-                .options(*self._get_eager_run_query_options()) \
-                .filter(
-                    SqlRun.experiment_id.in_(experiment_ids),
-                    SqlRun.lifecycle_stage.in_(stages),
-                    *_get_attributes_filtering_clauses(parsed_filters)) \
+            queried_runs = query \
+                .filter(SqlRun.experiment_id.in_(experiment_ids),
+                        SqlRun.lifecycle_stage.in_(stages),
+                        *_get_attributes_filtering_clauses(parsed_filters)) \
                 .order_by(*parsed_orderby) \
-                .offset(offset).limit(max_results).all()
-
-            runs = [run.to_mlflow_entity() for run in queried_runs]
+                .offset(offset).limit(max_results)
+
+            all_runs = queried_runs.all()
+
+            # equivalent to subquery load strategy but includes filter feature
+            # see https://groups.google.com/forum/m/#!msg/sqlalchemy/zr99U2onipY/6lAB0fy3sGsJ
+            # for explanation
+            all_runs = disjoint_load(queried_runs, all_runs, SqlRun.tags,
+                                     SqlTag.key.in_(tags_whitelist)
+                                     if tags_whitelist is not None else None)
+            all_runs = disjoint_load(queried_runs, all_runs, SqlRun.latest_metrics,
+                                     SqlLatestMetric.key.in_(metrics_whitelist)
+                                     if metrics_whitelist is not None
+                                     else None)
+            all_runs = disjoint_load(queried_runs, all_runs, SqlRun.params,
+                                     SqlParam.key.in_(params_whitelist)
+                                     if params_whitelist is not None else None)
```

Comment:
Can you explain how this logic works? Is this creating a SqlAlchemy query with all these criteria or is this going back to the database repeatedly and pulling more data?